Name: Location Sharing in Messaging
Version: 1.4.5-1
Author: Jason Robitaille
Description: All plain-text GPS coordinates sent/recieved will be encoded as links that will open to Google Maps. In addition, in the app menu will be an Insert GPS Location option.

--- .orig/usr/palm/applications/com.palm.app.messaging/app/controllers/chatview-assistant.js
+++ /usr/palm/applications/com.palm.app.messaging/app/controllers/chatview-assistant.js
@@ -67,6 +67,14 @@
 			command: 'addPicture',
 			disabled: false
 		};
+		
+		this.appMenuInsertCoordsItem = {
+			label: $L('Insert GPS Location'),
+			command: 'insertCoords',
+			disabled: false,
+			visible: true,
+			shortcut:'l'
+		};
 
 		this.appMenuBlockUsernameItem = {
 			label: $L('Block this person'),
@@ -94,6 +102,7 @@
 			items: [
 				chatEditItems,
 				this.appMenuAddPictureItem,
+				this.appMenuInsertCoordsItem,
 				Mojo.Menu.helpItem
 			]
 		};
@@ -328,6 +337,28 @@
 			if (event.command === 'addPicture') {
 				this.attachFilePicker();
 				event.stop();
+			} else if (event.command === 'insertCoords') {
+				this.controller.serviceRequest("palm://com.palm.location", {
+					method:"getCurrentPosition",
+					parameters: {
+						accuracy:1,
+						maximumAge :10,
+						responseTime:1
+					},
+					onSuccess: function(response) {
+						if(response.errorCode==0) {
+							var currText = this.messageTextWidget.mojo.getValue();
+							if(currText.length>0) {
+								if(currText.charAt(currText.length-1) != " ") {
+									currText += " ";
+								}
+							}
+							currText += response.latitude + "," + response.longitude + " ";
+							this.messageTextWidget.mojo.setValue(currText);
+						}
+					}.bind(this)
+				});
+				event.stop();
 			} else if (event.command === 'blockUsername') {
 				this.blockUsername();
 				event.stop();
@@ -1254,6 +1285,21 @@
 			(function()
 			{
 				Mojo.Log.info("********** RUNNING DELAYED TEXT INDEXER ***************");
+				var coordIndexer = function(text) {
+					var str = text;
+					var coordFilter = /-?\d+(\.?\d+)?,-?\d+(\.?\d+)?/g;
+					var result = "";
+					var matches = str.match(coordFilter);
+					while(matches) {
+						var matchIndex = str.indexOf(matches[0]);
+						result += str.substring(0, matchIndex);
+						result += "<a href=\"maploc:" + matches[0] + "\">" + matches[0] + "</a>";
+						str = str.substring(matchIndex + matches[0].length);
+						matches = str.match(coordFilter);
+					}
+					result += str;
+					return result;
+				}
 				var val = delayed.summary;
 				if (val)
 				{
@@ -1258,6 +1304,7 @@
 				if (val)
 				{
 					var nval = PalmSystem.runTextIndexer(val);
+					nval = coordIndexer(nval);
 					if (nval != val)
 					{
 						itemModel.summary = nval;
@@ -1268,6 +1315,7 @@
 				if (val)
 				{
 					var nval = PalmSystem.runTextIndexer(val);
+					nval = coordIndexer(nval);
 					if (nval != val)
 					{
 						itemModel.messageText = nval;
